src/app/ui/textedit.cpp

```diff
@@ -0,0 +1,748 @@
+    const auto& lineBounds = line.getBounds(m_selection.start().pos(),
+                                            m_lines[m_selection.start().line()].glyphCount);
+    selectionRect.x += lineBounds.x;
+    selectionRect.w = lineBounds.w;
+  }
+
+  selectionRect.h = line.height; // Normalize the height of the rect so it doesn't vary.
+  return selectionRect;
+}
+
+TextEdit::Caret TextEdit::caretFromPosition(const gfx::Point& position)
+{
+  const auto* view = View::getView(this);
+  if (!view)
+    return Caret();
+
+  if (m_lines.empty())
+    return Caret(&m_lines, 0, 0);
+
+  // Deduce the position the user wants to go when clicking outside of bounds
+  if (!view->viewportBounds().contains(position)) {
+    if (position.y < view->viewportBounds().y) {
+      return Caret(&m_lines, 0, 0);
+    }
+
+    if (position.y > view->viewportBounds().y2()) {
+      return Caret(&m_lines, m_lines.size() - 1, m_lines.back().glyphCount);
+    }
+
+    if (position.x > view->viewportBounds().x2()) {
+      Caret caret = m_caret;
+      caret.right();
+      return caret;
+    }
+
+    return Caret();
+  }
+
+  // Normalize the mouse position to the internal coordinates of the widget
+  gfx::PointF offsetPosition(position.x - (bounds().x + border().left()),
+                             position.y - (bounds().y + border().top()));
+
+  offsetPosition += View::getView(this)->viewScroll();
+
+  Caret caret(&m_lines);
+  const int lineHeight = font()->height();
+
+  // First check if the offset position is blank (below all the lines)
+  if (offsetPosition.y > m_lines.size() * lineHeight) {
+    // Get the last character in the last line.
+    caret.setLine(m_lines.size() - 1);
+
+    // Check the line width and if we're more than halfway past the line, we can set the caret to
+    // the full line.
+    caret.setPos(
+      (offsetPosition.x > m_lines[caret.line()].width / 2) ? m_lines[caret.line()].glyphCount : 0);
+    return caret;
+  }
+
+  for (const Line& line : m_lines) {
+    const size_t lineStartY = line.i * lineHeight;
+    const size_t lineEndY = (line.i + 1) * lineHeight;
+
+    if (offsetPosition.y > lineEndY || offsetPosition.y < lineStartY)
+      continue; // We're not in this line
+
+    caret.setLine(line.i);
+
+    if (!line.blob)
+      break; // Line has no text, we can end it here.
+
+    if (offsetPosition.x > line.width) {
+      // Clicking on the blank space next to a line should put our caret at the end of it.
+      caret.setPos(line.glyphCount);
+      break;
+    }
+
+    // Find the exact character we're standing on, with a slight bias to the left or right
+    // depending on where we click wrt the glyph bounds
+    size_t advance = 0;
+    bool found = false;
+
+    line.blob->visitRuns([&](const text::TextBlob::RunInfo& run) {
+      if (found) {
+        return;
+      }
+
+      for (int i = 0; i < run.glyphCount; ++i) {
+        const gfx::RectF& glyphBounds = run.getGlyphBounds(i).offset(gfx::PointF(0, lineStartY));
```

Comment:
This is not a full review and I'm not sure if it's working with MSVC, but with clang-18 on Linux the clicking/selecting a char with line>1 and column>1 didn't work, and just changing this to:
```suggestion
        gfx::RectF glyphBounds = run.getGlyphBounds(i).offset(gfx::PointF(0, lineStartY));
```
Will fix the selection in multiple lines/clicking specific chars in any line. Never use a reference to a temporal object :pray: